multi/stm32l4-multi/flash.c

```diff
@@ -232,6 +244,68 @@ size_t flash_writeData(uint32_t offset, const char *buff, size_t size)
 }
 
 
+static unsigned int _getOptions(void)
+{
+	return *(flash_common.flash + flash_optr);
+}
+
+
+static int _setOptions(unsigned int mask, unsigned int val)
+{
+	int err;
+	unsigned int t;
+
+	if ((err = _flash_wait()) < 0)
+		return err;
+
+	_flash_clearFlags();
+
+	t = *(flash_common.flash + flash_optr) & ~mask;
+	*(flash_common.flash + flash_optr) = t | (mask & val);
+	dataBarier();
+	*(flash_common.flash + flash_cr) |= 1 << 17;
+
+	err = _flash_wait();
+	_flash_clearFlags();
+
+	return err;
+}
+
+
+extern int flash_switchBanks(void)
+{
+	int err;
+	unsigned int bfb2;
+
+	mutexLock(flash_common.lock);
+	bfb2 = !!(_getOptions() & opt_bfb2);
+	_program_unlock();
+	_program_optunlock();
+
+	/* If BFB2==1 and Bank 1 is the active bank, then booting from Bank 2
```

Comment:
Wonder if it's possible that bfb2 == 0 and Bank2 is active bank (fail to boot from Bank 1) or it can never happen.

Reply:
According to ST, when bfb2 == 0 then the device boots directly from Bank 1 and does not perform any checks. Only when bfb2 == 1 the bootloader checks Bank 2's initial stack pointer and decides whether to boot from Bank 1 or Bank 2. So with bfb2 == 0 there is no fail-safe.